docs/proposals/multiple-sources-for-applications.md

```diff
@@ -0,0 +1,225 @@
+
+## Motivation
+
+The main motivation behind this enhancement proposal is to allow users to create an application using services that are stored in various sources.
+
+### Goals
+
+The goals of the enhancement are:
+
+#### **Supporting multiple sources for creating an application**
+
+Users should be able to specify multiple sources to add services to the application. Argo CD should compile all the sources and reconcile each source individually for creating the application.
+
+#### **Allow specifying external value files for Helm repositories**
+
+Users should be able to specify different sources for Helm charts and values files. The Helm charts specified by the user could be available in Git or Helm repository and the value files are stored in Git. Argo CD should track changes in both the Helm charts and the value files repository and reconcile the application.
+
+#### Changes to UI
+
+The UI should allow users to add multiple sources while creating the application. For each source, UI should allow users to add multiple external values files Helm projects.
+
+#### Changes to cli
+
+The cli would need to support adding a list of resources instead of just one while creating the application. Also, just like `--values` field for adding values files, we would need an option to allow external value files to the application.
+
+### Non-goals
+* Allow reconciliation from Argo CD Application resources that are located at various sources. We believe this would be possible with some adaptions in the reconcilation workflow.
+
+## Proposal
+
+### Add new `sources` field in Application Spec
+
+The proposal is to add a new field `sources` which would allow users to input list of `ApplicationSource`s. We would mark field `source` as deprecated and would override the details under `source` with details under `sources` field.
+
+```yaml
+spec:
+  source:
+    repoURL: https://charts.bitnami.com/bitnami
+    targetRevision: 8.5.8
+    helm:
+    valueFiles:
+        - values.yaml
+    chart: mysql
+  sources:                                          # new field
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+      chart: mysql
+```
+
+### Add `externalValues` field in helm section of Application Spec
+
+Along with new `sources` field, add a new field for accepting external value files `externalValueFiles` under the `helm` section of each ApplicationSource.
+
+```yaml
+sources:
+    - repoURL: https://charts.bitnami.com/bitnami
+      targetRevision: 8.5.8
+      helm:
+        valueFiles:
+          - values.yaml
+        externalValueFiles:                         # new field
+          - repoURL: https://github.com/KaiReichart/argo-test-values.git
+            targetRevision: main
+            valueFiles:
+              - values.yaml
```

Comment:
```suggestion
        valueFiles:
          - values.yaml
          - ../my-other-source/values.yaml
```

I like the `externalValueFiles` field. It's very clear/explicit about what it does. At the same time, I think we should consider whether we want to allow `source-a` to access files in `source-b`. The "Helm chart + values from git" use case is the only one I can think of for the moment. But I could imagine cross-source resource sharing being used for other features as well.

If we _don't_ prefer the `../my-other-source` approach, we should consider whether there are any security concerns around cross-source communication which would lead us to disable it entirely.

Reply:
```suggestion
        valueFiles:
          - values.yaml
        externalValueFiles:
          - sourceName: my-source
            path: override-values.yaml
```

If the sources were named, we could do something like this.

Reply:
I like the idea of cross-source resource sharing. But we would have to evaluate security concerns around it.